File: swift_src/scalar.swift
//
//  scalar.swift
//  Definition of multitude of optimal strategies
//
//  Created by Евгений Канашкин on 15.11.2024.
//

import Surge

// Результатом данной функции является скаляр

//MARK: - Генерация v (не инвертирует матрицу)
func getScalarV(_ matrix: Matrix<Double>) -> Double? {
        
        ifdebug(.scalar, .start, "Генерация скаляра матрицы.")
        
        var colSum: [Double] = []
        for row in 0 ..< matrix.rows {
                colSum.append(sum(Array(transpose(matrix)[row])))
        }
        
        ifdebug(.scalar, .info, "\(colSum) -> \(sum(colSum)).")
        
        if abs(sum(colSum)) < 1e-10 {
                ifdebug(.scalar, .warn, "Неудачная генерация скаляра. v == nil.")
                return nil
        }
        
        let result = 1 / sum(colSum)
        
        ifdebug(.scalar, .stop, "Генерация скаляра завершена. \(result).")
        return result
}

File: src/scalar.rs
use nalgebra::DMatrix;

use crate::{debug_info, Functions, IfType};

pub fn get_scalar_v(matrix: &DMatrix<f64>) -> Option<f64> {
    debug_info(Functions::Scalar, IfType::Start, "Генерация скаляра матрицы.");

    // Сумма по столбцам
    let mut col_sum: Vec<f64> = vec![0.0; matrix.ncols()]; // Инициализируем вектор для суммы столбцов

    for row in 0..matrix.nrows() {
        for col in 0..matrix.ncols() {
            col_sum[col] += matrix[(row, col)];
        }
    }

    let total_sum = col_sum.iter().sum::<f64>();
    debug_info(Functions::Scalar, IfType::Info, &format!("{:?} -> {}", col_sum, total_sum));

    if total_sum.abs() < 1e-10 {
        debug_info(Functions::Scalar, IfType::Warn, "Неудачная генерация скаляра. v == None.");
        return None; // Возвращаем None, если сумма близка к нулю
    }

    let result = 1.0 / total_sum;

    debug_info(Functions::Scalar, IfType::Stop, &format!("Генерация скаляра завершена. {}.", result));
    Some(result)
}